### PR TITLE
Enable Invalidation and Reloading via controller

### PR DESCRIPTION
**Pull request by JWambaugh**
Change to allow invalidation based on the discussion here: https://github.com/deakjahn/huge_listview/issues/7
Adds a controller that allows users to refresh the list and change the number of items in the list.

### CODE REVIEW

File: lib/src/huge_listview_controller.dart [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import 'package:flutter/material.dart';
+
+class HugeListViewController {
+  void Function()? onReload;
+  void Function(bool reloadPage)? onInvalidateList;
+  void Function(int)? setTotalItemCount;
+
+  set totalItemCount(int count) {
+    if (setTotalItemCount != null) {
```
Comment (deakjahn) [inline lib/src/huge_listview_controller.dart:9]: Not terribly important but I would normally use:

    if (callback == null) throw Exception("...");
    callback();

Less verbose and the compiler/IDE finds that `callback()` isn't null by itself, without `!`.

**Comment by deakjahn** [comment]
Nice but my only worry is that the example, now modified, no longer shows how to really use it with a completely dynamically generated list. It would be better to have two samples somehow, to keep both old and new, with and without controller.

Although it adds to the complexity, I can't really see a better way than to add another page besides `MyHomePage` to accommodate both. I wonder if we would need to worry about an UI to select or simply let people to call whichever they want to see. I'm leaning towards this latter...

**Comment by deakjahn** [comment]
@JWambaugh Now that I started to look into it in more detail, I start to have my doubts, sorry. The code works as it is, sure, but this isn't how we normally use controllers in Flutter. Controllers are used the other way around: they are change notifiers that notify their listeners of any incoming change. Instead of the view populating its callbacks, they subscribe to the controller's notifications. And for full compliance, the usual way is to _always_ use controllers: if the user doesn't provide one, creating its own so that it can delegate all those tasks to the controller all the time.

I have the feeling that we can easily paint ourselves into a corner if we don't follow the regular route when, the next time around, we will need to add even more functionality to the controller.

What I'd like to suggest now is to close this, I will use all the code and ideas contained here to come up with a properly architected controller (I have an analogous one in another project, no need to think about it even), upload it here (without publishing first, of course) and let's see what you think about that approach. Agreed?

**Comment by JWambaugh** [comment]
Sounds good @deakjahn. 
This wasn't my design but a quick implementation of what was discussed in #7 (because I need some way to refresh the list for my project) and I thought I'd share.

A more thought-out solution sounds good.

File: lib/src/huge_listview_controller.dart (hunk 1 above)
Comment (JWambaugh) [reply lib/src/huge_listview_controller.dart:9]: yep, that's better for sure.

**Comment by deakjahn** [comment]
OK, I already did the better part of it, no more syntax errors but I haven't yet started to test the example. But it's late into the night over here, so I'll have to postpone it until tomorrow. :-)

**Comment by deakjahn** [comment]
@JWambaugh Let's see now, https://github.com/deakjahn/huge_listview/commit/80a9563b33c6c3900420ae3cac3c67869a8b97cb.

The solution mostly stayed the same but the controller is set up differently and the related old approach is deprecated. Total item count has been moved to the controller completely (well, it stays in the view temporarily because of the deprecation but it shouldn't be used; I guess the controller will end up being required rather than optional in the end, anyway).

I thought long and hard about those callbacks and implemented a different approach using booleans in the value notifier. This is also kind of a hack but, I think, still much less so. Dart has no notion of package private access, so leaving two callbacks on the controller open for any user to reach and mess with just felt wrong (yes, I found the blog post that [3rpse-ai](https://github.com/3rpse-ai) also probably found that suggested this approach).

One thing I'm not yet sure about is the case of the `ValueListenableBuilder` you suggested around the `DraggableScrollbar`. Now that the operation is basically turned inside out, the controller pushing the notification rather than the view pulling it -- as it should be --, it isn't necessary. Still, you mentioned something about the smooth behavior of the scrollbar during reloads and I can't now decide whether we reached that already not.

The example now has two pages, the old example and the new one.

I also formatted the whole package now. I hate this 80-character-wide stupidity of dartfmt with passion (I never use it in my own code, I have a large monitor and long lines) but unfortunately, pub.dev enforces it. So, it looks like more changes than actually the case.

**Comment by deakjahn** [comment]
@JWambaugh Check it out, please, when you'll have time...

**Comment by JWambaugh** [comment]
[deakjahn](https://github.com/deakjahn) Looks great, and looks to be working great too! thank you!

**Comment by deakjahn** [comment]
@JWambaugh OK, then let's publish and hope for the best...

**Pull request closed without merging**